pkg/azure/aks/pricing_map.go

```diff
@@ -0,0 +1,183 @@
+	logger            *slog.Logger
+	context           context.Context
+	retailPriceClient *retailPriceSdk.RetailPricesClient
+
+	RegionMap map[string]PriceByPriority
+	Cache     map[string]*retailPriceSdk.ResourceSKU
+}
+
+func NewPricingStore(subId string, priceClient *retailPriceSdk.RetailPricesClient, parentLogger *slog.Logger, parentContext context.Context) *PriceStore {
+	logger := parentLogger.With("subsystem", "pricingMap")
+
+	p := &PriceStore{
+		lock:              &sync.RWMutex{},
+		logger:            logger,
+		context:           parentContext,
+		subscriptionId:    subId,
+		retailPriceClient: priceClient,
+
+		RegionMap: make(map[string]PriceByPriority),
+		Cache:     make(map[string]*retailPriceSdk.ResourceSKU),
+	}
+
+	go func() {
+		err := p.PopulatePriceStore([]string{})
+		if err != nil {
+			p.logger.LogAttrs(p.context, slog.LevelError, "error populating initial price store", slog.String("error", err.Error()))
+		}
+	}()
+
+	return p
+}
+
+func (p *PriceStore) buildQueryFilter(locationList []string) string {
+	if len(locationList) == 0 {
+		return `serviceName eq 'Virtual Machines' and priceType eq 'Consumption'`
+	}
+
+	locationListFilter := []string{}
+	for _, region := range locationList {
+		locationListFilter = append(locationListFilter, fmt.Sprintf("armRegionName eq '%s'", region))
+	}
+
+	locationListStr := strings.Join(locationListFilter, " or ")
+	return fmt.Sprintf(`serviceName eq 'Virtual Machines' and priceType eq 'Consumption' and (%s)`, locationListStr)
+}
+
+func (p *PriceStore) buildListOptions(locationList []string) *retailPriceSdk.RetailPricesClientListOptions {
+	queryFilter := p.buildQueryFilter(locationList)
+	return &retailPriceSdk.RetailPricesClientListOptions{
+		APIVersion:  to.StringPtr(AZ_API_VERSION),
+		Filter:      to.StringPtr(queryFilter),
+		MeterRegion: to.StringPtr(`'primary'`),
+	}
+}
+
+func (p *PriceStore) determineMachineOperatingSystem(sku retailPriceSdk.ResourceSKU) MachineOperatingSystem {
+	switch {
+	case strings.Contains(sku.ProductName, "Windows"):
+		return Windows
+	default:
+		return Linux
+	}
+}
+
+func (p *PriceStore) determineMachinePriority(sku retailPriceSdk.ResourceSKU) MachinePriority {
+	switch {
+	case strings.Contains(sku.SkuName, "Spot"):
+		return Spot
+	default:
+		return OnDemand
+	}
+
+}
+
+func (p *PriceStore) PopulatePriceStore(locationList []string) error {
+	startTime := time.Now()
+	p.logger.LogAttrs(p.context, slog.LevelInfo, "populating price map")
+
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	pager := p.retailPriceClient.NewListPager(p.buildListOptions(locationList))
+
+	for pager.More() {
+		page, err := pager.NextPage(p.context)
+		if err != nil {
+			p.logger.LogAttrs(p.context, slog.LevelError, "error paging")
+			return ErrPageAdvanceFailure
+		}
+
+		for _, v := range page.Items {
+			regionName := v.ArmRegionName
+			if regionName == "" {
+				p.logger.LogAttrs(p.context, slog.LevelInfo, "region name for price not found")
+				continue
```

Comment:
Is there a specific example where this occurs? It might be worthwhile to link out to a specific sku if possible.

Reply:
sure - added